SARAALERT-1240: Update patient updated_at upon modification of associated records

timwongj opened the pull request:
# Description
Jira Ticket: [SARAALERT-1240](https://tracker.codev.mitre.org/browse/SARAALERT-1240)

Adding/updating any associated records (labs, close contacts, histories, etc) with a Patient should prevent the patient from being purge eligible

# Important Changes
Please list important files (meaning substantial or integral to the PR) along with a list of the general changes that should be highlighted for reviewers.

`example_file.js`
- Example change (ex: refactored import function)

# Testing
This fix was tested on the following browsers (submitter must check all those that apply):
* [x] Chrome
* [ ] Firefox
* [ ] Safari
* [ ] IE11

Please describe the tests needed to verify your changes. Provide instructions for reproducing these tests. List any relevant details for your test configuration.

Bialogs reviewed (changes requested):
Do you need `after_save` and `before_destroy`?

Calling touch after_save would just update the timestamp again on `self`, right? 

Not sure why we would need to update the timestamp for `self` right before we destroy the object? 

I also removed these lines and re-ran the tests you wrote and they pass.

https://guides.rubyonrails.org/association_basics.html#options-for-belongs-to-touch

timwongj commented:
> Do you need `after_save` and `before_destroy`?
> 
> Calling touch after_save would just update the timestamp again on `self`, right?
> 
> Not sure why we would need to update the timestamp for `self` right before we destroy the object?
> 
> I also removed these lines and re-ran the tests you wrote and they pass.
> 
> https://guides.rubyonrails.org/association_basics.html#options-for-belongs-to-touch

Thanks for the catch, I think I might have been testing it incorrectly when I was writing it at first and thought `patient.touch` might only be called if `assessment.touch` was explicitly called, but glad it works without it. Fixed here: https://github.com/SaraAlert/SaraAlert/pull/683/commits/c17cbe274d6efe5cc23017f9bbb6005ae18b0e39

kylietmo commented on test/models/assessment_test.rb:
```diff
@@ -8,6 +8,28 @@ def setup; end
+    patient.update(updated_at: 1.day.ago)
+    ActiveRecord::Base.record_timestamps = true
+    assessment = create(:assessment, patient: patient)
+    assert_in_delta patient.updated_at, DateTime.now, 1
```
I think this delta is too large. It's a delta of an entire day when if anything it would be a matter of MS difference I think.

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/683/commits/dce4d4b479a47f020b4151fc583a54debffeae6e

kylietmo reviewed (changes requested):
You should also add a `touch: true` and accompanying tests to `ContactAttempt` here: https://github.com/SaraAlert/SaraAlert/blob/master/app/models/contact_attempt.rb

kylietmo reviewed (changes requested):
The ticket has been updated with additional criteria for this for History items. Specifically:
> ANY history item that is about sending the monitoree a message via the system automatically (whether it was a success or failure) should NOT be considered a record update. Nor should downloading/exporting that record data.

kylietmo commented on app/models/history.rb:
```diff
@@ -39,6 +39,13 @@ class History < ApplicationRecord
+  before_destroy :touch_patient
+
+  def touch_patient
+    patient.touch unless history_type == HISTORY_TYPES[:report_reminder]
```
This is difficult but we seem to use various History types for this kind of thing which is a consistency issue. In both the `PatientMailer` and the `ConsumeAssessmentJobs, we update a patient record with history items of type `Contact Attempt` if there's an issue as well which we wouldn't want to be included as a touch based on our criteria. It doesn't look like we use `Contact Attempt` history types for anything the user does so many this just needs to be updated to check both `report_reminder` and `contact_attempt` history types.

timwongj replied:
I think we have to account for manual contact attempts:

![Screen Shot 2021-02-10 at 1 59 29 PM](https://user-images.githubusercontent.com/9956561/107557943-36066a80-6ba8-11eb-9cbb-3da97a29774a.png)

so I just checked for `created_by` to be `Sara Alert System`: https://github.com/SaraAlert/SaraAlert/pull/683/commits/7bc9ddfc5b56b8c866b3e3201e438225a24315b6

kylietmo replied:
Good point. Would you mind adding a comment around that line that explains the logic?

timwongj replied:
sure! https://github.com/SaraAlert/SaraAlert/pull/683/commits/f08bac4ad3811e049e75476f487dacacf231f40f

Bialogs replied:
Isn't this still missing the export case?

timwongj replied:
@Bialogs are you talking about the `:monitoree_data_download` history_type?

kylietmo replied:
Right good catch @Bialogs - I think we need to exclude that history type as well @timwongj

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/683/commits/1987337042767c09797d9cb63d699f585d003f30

kylietmo replied:
Thanks, we should also have tests for these cases specifically in the related test file

timwongj replied:
https://github.com/SaraAlert/SaraAlert/pull/683/commits/1abe7f833a84aaf78e5f96a5cd1aeca53fcb80c3

kylietmo commented:
Could you please change the target of this to be `1.24.0`?

holmesie commented on app/models/history.rb:
```diff
@@ -39,6 +39,15 @@ class History < ApplicationRecord
   belongs_to :patient
 
+  after_save :touch_patient
+  before_destroy :touch_patient
```
I think this could just be after_create.

timwongj replied:
addressed here: https://github.com/SaraAlert/SaraAlert/pull/683/commits/1987337042767c09797d9cb63d699f585d003f30

holmesie commented on app/models/history.rb:
```diff
@@ -39,6 +39,15 @@ class History < ApplicationRecord
+
+  def touch_patient
+    # patient updated_at should not be updated if history was created by a system generated report reminder or contact attempt
+    patient.touch unless history_type == HISTORY_TYPES[:report_reminder] ||
```
I think this should be decoupled a bit. Why not introduce a `is_user_driven?` method on this model or something (horrible name I know, but you get the idea). The method would tell you if a History instance was user driven or not.

You could probably even eliminate this whole method, and use a nice and succinct lambda on the callback above.

Bialogs replied:
I was thinking the exact same thing. I wrote something that tried to go back to adding an if condition to `:touch` itself which I don't think is possible so I didn't mention it. But if the whole thing is just an `after_create` then the logic can come in.

holmesie replied:
I too went down that "can I use a lambda or a proc with touch?" and sadly was disappointed...

timwongj replied:
yeah that's what I tried to do as well and wasn't able to find any way to do a conditional touch, are we still able to simplify this by introducing a `user_driven?` method?

holmesie replied:
https://guides.rubyonrails.org/active_record_callbacks.html#using-if-and-unless-with-a-proc is what I was thinking

holmesie replied:
This is what I was thinking: https://en.wikibooks.org/wiki/Ruby_on_Rails/ActiveRecord/Callbacks#Proc

timwongj replied:
how's this? https://github.com/SaraAlert/SaraAlert/pull/683/commits/792a5ecd740a4f5aa5729400b0edd21b90940a9a

holmesie replied:
Move the logic to its own method. That lambda should only be one line. Also really should figure out how to not need to ignore any rubocop complaints.

Sorry to be nit-picky, but this is the first time we've done something like this for this app - and it's important to establish a good pattern for others to follow moving forward.

timwongj replied:
no worries, I'm just a bit unsure how exactly this should be factored out as its own method. wouldn't doing that essentially revert it back to how it was before (`touch_patient`)? or did you mean something like this?

```
  # patient updated_at should not be updated if history was created by a system generated action
  after_create proc { |history| history.patient.touch unless system_generated_action? }

  def system_generated_action?
    history_type == HISTORY_TYPES[:report_reminder] || history_type == HISTORY_TYPES[:monitoree_data_downloaded] ||
    (history_type == HISTORY_TYPES[:contact_attempt] && created_by == 'Sara Alert System')
  end
```

timwongj replied:
https://github.com/SaraAlert/SaraAlert/pull/683/commits/f2b282b88dc4c87fefd179875eb1f7157ac72b18

holmesie commented on app/models/history.rb:
```diff
@@ -39,12 +39,12 @@ class History < ApplicationRecord
   belongs_to :patient
 
-  after_create :touch_patient
+  # patient updated_at should not be updated if history was created by a system generated action
```
Capitalize `patient` in this comment.

timwongj replied:
fixed

holmesie commented on app/models/history.rb:
```diff
@@ -39,12 +39,12 @@ class History < ApplicationRecord
 
-  after_create :touch_patient
+  # patient updated_at should not be updated if history was created by a system generated action
+  after_create(proc { |history| history.patient.touch unless system_generated_action? })
```
what about `after_create Proc.new { |history| history.patient.touch unless system_generated_action? }`

holmesie replied:
These two things seem inconsistent: `history.patient.touch` vs. `system_generated_action?` (should `system_generated_action?` be `history.system_generated_action?`?)

timwongj replied:
that's what I had originally but rubocop was being extremely picky about how it wanted the syntax to be, as for the inconsistency, I just tried it out and it seems like any combination of it works, so I just changed it to:
`after_create(proc { patient.touch unless system_generated_action? })`

kylietmo replied:
Sorry additional nitpick that isn't a big deal: `system_generated_action` is a bit misleading given the download is not system generated, and there are technically other system generated actions that don't trigger this. Maybe like `system_report_action_or_download?` That's not great either I'm just spitballing

timwongj replied:
updated!

Bialogs reviewed (approved):
Updated approval with new changes from `history.rb`

kylietmo commented:
@timwongj Base branch was changed to `1.24.0`. Can you rebase off of `1.24.0`?

kylietmo commented on app/models/history.rb:
```diff
@@ -39,6 +39,14 @@ class History < ApplicationRecord
 
   belongs_to :patient
 
+  # Patient updated_at should not be updated if history was created by a system generated action
```
Might wanna clarify this comment - it's very specific cases we don't want to update the Patient record, not just any system generated action. The cases are 1) user collecting data about the monitoree and 2) the system still trying to send messages to the record

kylietmo replied:
I actually just realized something very silly about what we had suggested early when re-reading this. Simply preventing the history messages related to sending messages is of course not enough to handle this use case:
> 2) the system still trying to send messages to the record 

The system won't be sending messages to records that are closed anyway, and the close logic does not look at `updated_at`, only the purge logic does. So I don't actually think we even need to check the `report_reminder` and `contact_attempt` history types here. Even if a message was somehow sent to a closed record (like in the HoH case), `last_assessment_reminder_sent` would be updated on the Patient record so this check doesn't really do anything. 

So I'm basically proposing we change this check entirely to only filter out the `monitoree_data_downloaded` history type. @holmesie thoughts?

Bialogs replied:
The original statement:
>>ANY history item that is about sending the monitoree a message via the system automatically (whether it was a success or failure) should NOT be considered a record update.

Does not have anything to do with the state the Patient is in. I think the logic here is still valid - do not touch the `updated_at` of a patient when history items about system-generated contact attempts and data exports are created.

I'm not sure where the requirement for these specific cases came from but it would be against the above request to leave out these cases just because the features, in the way that they are currently implemented, do not take this into account.

Unless of course that the request wasn't discussing in general, but just as a way to have the "correct" Patient show up for purge eligible.

kylietmo replied:
>I'm not sure where the requirement for these specific cases came from but it would be against the above request to leave out these cases just because the features, in the way that they are currently implemented, do not take this into account.

Requirements came from discussion with PO team where we were revisiting the close/purge logic and this came up. 
We are essentially looking into updating likely the close logic to check for very stale records in the system - in which case we may be checking the `updated_at` field, and we didn't want that to be effected by automated messages or user downloads.

You're right in that we shouldn't be focused on the current close logic (because if we update the logic as I mentioned above that won't matter), so the more important part of what I said above is that when we do send messages, we already update an attribute on the Patient record (`last_assessment_reminder_sent`) so this change is rendered useless. I'm basically saying, because of this, we'd need another way to check for stale records other than looking at the `updated_at` anyway, so I don't think we need these extra checks.

timwongj replied:
sounds good, so is the verdict to remove the `:report_reminder` and `:contact_attempt` edge cases and just keep the `:monitoree_data_download` one for the scope of this ticket and then update the purge logic in a separate ticket or try to do that here?

kylietmo replied:
Yes that's what I'm thinking. Would like @holmesie thought's before you make changes though.

holmesie replied:
That makes sense to me as well @timwongj

Bialogs replied:
@kylietmo thanks for the additional explanation

timwongj replied:
updated here: https://github.com/SaraAlert/SaraAlert/pull/683/commits/6a640aa26e785d5c839ca18884a3ee37181ea4ef

kylietmo reviewed (approved):
Looks good to me. I just had one small request about a comment but not a blocker.

kylietmo commented on test/models/history_test.rb:
```diff
@@ -42,6 +42,22 @@ def history_types
     end
   end
 
+  test 'update patient updated_at upon history create, update, and delete' do
```
Test name should be updated (we don't check update/delete for History).

timwongj replied:
fixed

kylietmo reviewed (approved):
Looks good to me. One small comment on test name. Please squash the commits when you get a chance.

timwongj commented:
> Looks good to me. One small comment on test name. Please squash the commits when you get a chance.

squashed
